compiler/semantic/expr.go

```diff
@@ -116,6 +116,42 @@ func (a *analyzer) semExpr(e ast.Expr) (dag.Expr, error) {
 			Name: "cast",
 			Args: []dag.Expr{expr, typ},
 		}, nil
+	case *ast.IndexExpr:
+		expr, err := a.semExpr(e.Expr)
+		if err != nil {
+			return nil, err
+		}
+		// XXX This should be type checked as either an int, string or var.
+		index, err := a.semExpr(e.Index)
+		if err != nil {
+			return nil, err
+		}
+		// If container is a path and the index expression is a string, then
+		// just extend the path.
+		if path := a.isIndexOfThis(expr, index); path != nil {
+			return path, nil
+		}
+		return &dag.IndexExpr{Kind: "IndexExpr", Expr: expr, Index: index}, nil
+	case *ast.SliceExpr:
+		expr, err := a.semExpr(e.Expr)
+		if err != nil {
+			return nil, err
+		}
+		// XXX Low/High should be type checked as either int or var.
```

Comment:
Since slice indices can be arbitrary expressions, I think we can only check their types if they're literals.
```suggestion
		// Literal indices should be type checked as int.
```